text/0617-cloudfront-oac-l2.md

```diff
@@ -0,0 +1,877 @@
+              },
+            },
+          },
+        );
+        const result = this.bucket.addToResourcePolicy(oacReadOnlyBucketPolicyStatement);
+        return result;
+      }
+
+      /**
+       * Use custom resource to update bucket policy
+       */
+      private grantDistributionAccessToImportedBucket(scope: Construct, distributionId: string) {
+        const provider = S3OriginAccessControlBucketPolicyProvider.getOrCreateProvider(scope, S3_ORIGIN_ACCESS_CONTROL_BUCKET_RESOURCE_TYPE,
+          {
+            description: 'Lambda function that updates S3 bucket policy to allow CloudFront distribution access.',
+          });
+        provider.addToRolePolicy({
+          Action: ['s3:getBucketPolicy', 's3:putBucketPolicy'],
+          Effect: 'Allow',
+          Resource: [this.bucket.bucketArn],
+        });
+
+        new CustomResource(scope, 'S3OriginBucketPolicyCustomResource', {
+          resourceType: S3_ORIGIN_ACCESS_CONTROL_BUCKET_RESOURCE_TYPE,
+          serviceToken: provider.serviceToken,
+          properties: {
+            DistributionId: distributionId,
+            AccountId: this.bucket.env.account,
+            Partition: Stack.of(scope).partition,
+            BucketName: this.bucket.bucketName,
+          },
+        });
+      }
+
+      /**
+       * Use custom resource to update KMS key policy
+       */
+      private grantDistributionAccessToKey(scope: Construct, distributionId: string, key: IKey) {
+        const provider = S3OriginAccessControlKeyPolicyProvider.getOrCreateProvider(scope, S3_ORIGIN_ACCESS_CONTROL_KEY_RESOURCE_TYPE,
+          {
+            description: 'Lambda function that updates SSE-KMS key policy to allow CloudFront distribution access.',
+          });
+        provider.addToRolePolicy({
+          Action: ['kms:PutKeyPolicy', 'kms:GetKeyPolicy', 'kms:DescribeKey'],
+          Effect: 'Allow',
+          Resource: [key.keyArn],
+        });
+
+        new CustomResource(scope, 'S3OriginKMSKeyPolicyCustomResource', {
+          resourceType: S3_ORIGIN_ACCESS_CONTROL_KEY_RESOURCE_TYPE,
+          serviceToken: provider.serviceToken,
+          properties: {
+            DistributionId: distributionId,
+            KmsKeyId: key.keyId,
+            AccountId: this.bucket.env.account,
+            Partition: Stack.of(scope).partition,
+          },
+        });
+      }
+    });
+  }
+
+  protected constructor(protected readonly bucket: s3.IBucket, props: S3OriginProps = {}) {
+    super(bucket.bucketRegionalDomainName, props);
+  }
+
+  public abstract bind(scope: Construct, options: cloudfront.OriginBindOptions): cloudfront.OriginBindConfig;
+
+  protected abstract renderS3OriginConfig(): cloudfront.CfnDistribution.S3OriginConfigProperty | undefined;
+
+  protected _bind(scope: Construct, options: cloudfront.OriginBindOptions): cloudfront.OriginBindConfig {
+    return super.bind(scope, options);
+  }
+}
+```
+
+#### `S3Origin` Construct Modifications
```

Comment:
Kill it with fire?

Reply:
OAC and OAI feel like two different things, so I think it's worth killing this as is.

Reply:
Agreed, they each have their own binding implementation so it makes sense to separate them out rather than add to the existing class.